### ISSUE TITLE
`cloneIfObject` converts arrays to objects

### ISSUE THREAD

**Issue by twisty** [formsy/formsy-react#385]
If you pass an array into this function, it returns an object.

https://github.com/formsy/formsy-react/blob/6fcc682e1dae82dd5643a744f587713f4d8a9ced/src/utils.ts#L47-L50

This is because this function returns `true` when passed an array:

https://github.com/formsy/formsy-react/blob/6fcc682e1dae82dd5643a744f587713f4d8a9ced/src/utils.ts#L7-L9

(Try in browser console)

```
typeof []
> "object"
```

I'm not certain if arrays need to be cloned too, but a workaround is:

```ts
export function cloneIfObject(value: unknown) {
  // Clone objects to avoid accidental param reassignment
  if (isArray(value)) {
    return [...value];
  }
  return isObject(value) ? { ...value } : value;
}
```

...but then `cloneIfObject` isn't very well named!

**Comment by twisty** [comment]
Maybe `cloneIfMutable` would be a better name?

**Comment by rkuykendall** [comment]
@twisty Do you think this 'feature' of immutability is even worth keeping around considering this issue and https://github.com/formsy/formsy-react/issues/379 ?

**Comment by rkuykendall** [comment]
It seems like what we want for this check is less like `isObject` which returns true for arrays?!?! and more like `isPlainObject` which would also fix the moment bug. https://github.com/lodash/lodash/blob/aa1d7d870d9cf84842ee23ff485fd24abf0ed3d1/isPlainObject.js#L30

**Comment by twisty** [comment]
`isPlainObject` makes sense.
 
I would have no problem if the cloning was ditched. It would then just be up to the user to handle immutability themselves when they call `setValue`.

If we do keep the cloning, we should deal with the arrays properly, though!

**Comment by rkuykendall** [comment]
Released:

# v2.0.1
- Respect dot notation in reset (#393)  a2d2e5f
- Remove unnecesary externalError state (#398)  d4eb73e
- Allow validation errors to be React elements (#394)  4096296
- Update rollup plugin libraries (#397)  bebefbe
- Improvements to README and PR Template (#395)  df5af84
- Fix types in Formsy spec tests (#396)  954370a
- Remove unused Formsy props; tighten types; improve runRules (#392)  0eaeef7
- Rename cloneIfObject to protectAgainstParamReassignment; Fix isObject (#391)  49fc634
- Added possibility to use browser builtin submit (#226)  761b3ef
- Improve test coverage (#378)  b8881d0
- Dependabot Automatic Library Bumps (#376)  8b5a0db
- Use React's New Context API (#198)  a79eec2
- Fix in-page link to updateInputsWithValue  6fcc682
- Change coveralls from dependency to dev dependency  c9d9db5
- Add coverage to gitignore  0afef60

https://github.com/formsy/formsy-react/compare/v2.0.0...v2.0.1
